training/CHAPEL_TUTORIAL.md

```diff
@@ -993,358 +1073,635 @@ Expected Output:
+var dst: [dom] int;
 
 resetCommDiagnostics();
 startCommDiagnostics();
-t1 = Time.timeSinceEpoch().totalSeconds();
+var t1 = Time.timeSinceEpoch().totalSeconds();
 
-// use refs to avoid repeated accesses
-ref ua = upperEntry.a;
-ref la = lowerEntry.a;
-forall (i, (up, low)) in zip(hashes.domain, hashes) {
-  ua[i] = up;
-  la[i] = low;
+forall (s, i) in zip(src, 0..) {
+  dst[ (i + (size / 2):int ) % size ] = s;
 }
 
-t2 = Time.timeSinceEpoch().totalSeconds() - t1;
+var t2 = Time.timeSinceEpoch().totalSeconds() - t1;
 stopCommDiagnostics();
-writeln("using refs time = ", t2);
-writeln();
+writeln("copy without aggregation time = ", t2);
+writeln("communication without aggregation: ");
 printCommDiagnosticsTable();
 
-
-var upper: [distDom] uint;
-var lower: [distDom] uint;
 resetCommDiagnostics();
 startCommDiagnostics();
 t1 = Time.timeSinceEpoch().totalSeconds();
 
-// iterate over arrays directly:
-// since they are distributed the same way,
-// the looping variables will always be local to each other
-forall (up, low, h) in zip(upper, lower, hashes) {
-  (up, low) = h;
+forall (s, i) in zip(src, 0..) with (var agg = new DstAggregator(int)) {
+  agg.copy(dst[ (i + (size / 2):int ) % size ], s);
 }
 
 t2 = Time.timeSinceEpoch().totalSeconds() - t1;
 stopCommDiagnostics();
-writeln("looping over arrays directly time = ", t2);
 writeln();
+writeln("copy with aggregation time = ", t2);
+writeln("communication using aggregation: ");
 printCommDiagnosticsTable();
 
-upperEntry = new SymEntry(upper);
-lowerEntry = new SymEntry(lower);
+if compareBulkTransfer {
+  resetCommDiagnostics();
+  startCommDiagnostics();
+  var t3 = Time.timeSinceEpoch().totalSeconds();
+  
+  // using aggregation is not actually needed
+  // since we are copying a contiguous block 
+  dst[0..<(size / 2)] = src[(size / 2)..<size];
+  dst[(size / 2)..<size] = src[0..<(size / 2)];
+  
+  var t4 = Time.timeSinceEpoch().totalSeconds() - t3;
+  stopCommDiagnostics();
+  writeln();
+  writeln("copy with aggregation time = ", t4);
+  writeln("communication using aggregation: ");
+  printCommDiagnosticsTable();
+}
 ```
 
-```
-leading iterator not distributed time = 1.00893
+```console
+$ chpl tutorial.chpl --no-cache-remote
+$ ./tutorial -nl 4
+copy without aggregation time = 13.2172
+communication without aggregation:
+| locale | get |    put | execute_on_nb |
+| -----: | --: | -----: | ------------: |
+|      0 |   7 | 250000 |             3 |
+|      1 |   7 | 250000 |             0 |
+|      2 |   7 | 250000 |             0 |
+|      3 |   7 | 250000 |             0 |
+
+copy with aggregation time = 0.159691
+communication using aggregation:
+| locale | get | put | execute_on | execute_on_nb |
+| -----: | --: | --: | ---------: | ------------: |
+|      0 |  80 |  40 |         40 |             3 |
+|      1 |  80 |  40 |         40 |             0 |
+|      2 |  80 |  40 |         40 |             0 |
+|      3 |  80 |  40 |         40 |             0 |
```

Comment:
Do you have any interest in defining what these different categories are?  It's a bit of a rathole, but also natural to wonder about.  My high-level explanation would be:

* "get": reading a value from a remote locale's memory
* "put": storing a value into a remote locale's memory
* "execute_on": creating a task on a remote locale, for example via an `on`-clause
* "_nb": non-blocking (doesn't prevent the original task from continuing)